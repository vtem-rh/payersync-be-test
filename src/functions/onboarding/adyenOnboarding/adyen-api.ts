import https from 'https';
import { URL } from 'url';

// --- API Client Creation ---
export interface AdyenApiClientConfig {
  apiKey: string;
  apiUrl: string;
}

export interface AdyenApiClients {
  lem: AdyenApiClientConfig;
  bp: AdyenApiClientConfig;
  psp: AdyenApiClientConfig;
}

export function createAdyenApiClients(
  lemApiKey: string,
  bpApiKey: string,
  pspApiKey: string,
  bpApiUrl: string,
  managementApiUrl: string,
  lemApiUrl: string
): AdyenApiClients {
  return {
    lem: { apiKey: lemApiKey, apiUrl: lemApiUrl },
    bp: { apiKey: bpApiKey, apiUrl: bpApiUrl },
    psp: { apiKey: pspApiKey, apiUrl: managementApiUrl },
  };
}

// --- API Call Functions ---

function handleAdyenRequest<T>(
  clientConfig: AdyenApiClientConfig,
  method: 'POST' | 'GET' | 'PATCH',
  path: string,
  payload: any,
  operation: string
): Promise<T> {
  const postData = JSON.stringify(payload);
  const url = new URL(clientConfig.apiUrl);

  // Ensure path is properly joined with the base URL
  // If the base URL already includes a path (like /lem/v4), we need to preserve it
  let finalPath;
  if (path.startsWith('/')) {
    // If path starts with /, ensure we don't lose the base path
    const basePathname = url.pathname.endsWith('/') ? url.pathname.slice(0, -1) : url.pathname;
    finalPath = `${basePathname}${path}`;
  } else {
    // If path doesn't start with /, just append it to the base path
    finalPath = url.pathname.endsWith('/') ? `${url.pathname}${path}` : `${url.pathname}/${path}`;
  }

  const options = {
    hostname: url.hostname,
    port: 443,
    path: finalPath,
    method: method,
    headers: {
      'Content-Type': 'application/json',
      'Accept': 'application/json',
      'X-API-Key': clientConfig.apiKey.trim(),
      'Content-Length': Buffer.byteLength(postData),
    },
    agent: false,
    timeout: 30000, // 30 seconds timeout
  };

  console.log(`Adyen ${operation} request sent to: ${clientConfig.apiUrl}${path}`, {
    headers: options.headers,
  });

  return new Promise<T>((resolve, reject) => {
    const req = https.request(options, res => {
      let data = '';
      res.on('data', chunk => {
        data += chunk;
      });
      res.on('end', () => {
        try {
          // Check if the response is likely HTML (common for error pages)
          const isHtmlResponse = data.trim().startsWith('<!DOCTYPE') || data.trim().startsWith('<html');

          if (isHtmlResponse) {
            console.error(`Adyen ${operation} returned HTML instead of JSON. Status: ${res.statusCode}`);
            // Include a snippet of the HTML in the error for debugging
            const htmlSnippet = data.length > 100 ? data.substring(0, 100) + '...' : data;
            reject(new Error(`Failed to ${operation}: Received HTML response instead of JSON. Status: ${res.statusCode}. Response starts with: ${htmlSnippet}`));
            return;
          }

          const responseData = data ? JSON.parse(data) : {};
          if (res.statusCode && res.statusCode >= 200 && res.statusCode < 300) {
            console.log(`Adyen ${operation} response - responseData:`, JSON.stringify(responseData, null, 2));
            resolve(responseData);
          } else {
            console.error(`Adyen ${operation} error response -n0 - responseData:`, data);
            reject(new Error(`Failed to ${operation}: ${responseData.detail || `Request failed with status code ${res.statusCode}`}`));
          }
        } catch (error) {
           console.error(`Adyen ${operation} parsing error:`, error);
           // Include the status code and full response for better debugging
           reject(new Error(`Failed to parse response for ${operation}. Status: ${res.statusCode}, Response: ${data}`));
        }
      });
    });

    req.on('error', error => {
      console.error(`Adyen ${operation} request error:`, error);
      reject(new Error(`Failed to ${operation}: ${error.message}`));
    });

    req.on('timeout', () => {
      console.error(`Adyen ${operation} request timeout after ${options.timeout}ms`);
      req.abort();
      reject(new Error(`Timeout: ${operation} request exceeded ${options.timeout}ms`));
    });

    req.write(postData);
    req.end();
  });
}

export async function createLegalEntity(lemClientConfig: AdyenApiClientConfig, legalEntityData: any): Promise<{ id: string }> {
  const path = '/legalEntities';
  return handleAdyenRequest(lemClientConfig, 'POST', path, legalEntityData, 'create legal entity');
}

export async function getLegalEntity(lemClientConfig: AdyenApiClientConfig, legalEntityId: string): Promise<any> {
  const path = `/legalEntities/${legalEntityId}`;
  return handleAdyenRequest(lemClientConfig, 'GET', path, null, 'get legal entity');
}

export async function createAccountHolder(bpClientConfig: AdyenApiClientConfig, accountHolderData: any, legalEntityId: string): Promise<{ id: string }> {
  const payload = { ...accountHolderData, legalEntityId };
  const path = '/accountHolders';
  return handleAdyenRequest(bpClientConfig, 'POST', path, payload, 'create account holder');
}

export async function createBalanceAccount(bpClientConfig: AdyenApiClientConfig, balanceAccountData: any, accountHolderId: string): Promise<{ id: string }> {
  const payload = { ...balanceAccountData, accountHolderId };
  const path = '/balanceAccounts';
  return handleAdyenRequest(bpClientConfig, 'POST', path, payload, 'create balance account');
}

export async function createBusinessLine(lemClientConfig: AdyenApiClientConfig, businessLineData: any, legalEntityId: string): Promise<{ id: string }> {
  const payload = { ...businessLineData, legalEntityId };
  const path = '/businessLines';
  return handleAdyenRequest(lemClientConfig, 'POST', path, payload, 'create business line');
}

export async function createStore(pspClientConfig: AdyenApiClientConfig, storeData: any, businessLineId: string, merchantId: string): Promise<{ id: string }> {
  const payload = { ...storeData, merchantId, businessLineIds: [businessLineId] };
  const path = '/stores';
  return handleAdyenRequest(pspClientConfig, 'POST', path, payload, 'create store');
}

export async function createOnboardingLink(lemClientConfig: AdyenApiClientConfig, hostedOnboardingData: any, legalEntityId: string): Promise<{ url: string }> {
  const { themeId, redirectUrl, locale, settings } = hostedOnboardingData;
  const payload = { themeId, redirectUrl, locale, settings };
  const path = `/legalEntities/${legalEntityId}/onboardingLinks`;
  return handleAdyenRequest(lemClientConfig, 'POST', path, payload, 'create onboarding link');
}

export async function createSplitConfiguration(pspClientConfig: AdyenApiClientConfig, merchantId: string): Promise<{ splitConfigurationId: string }> {
  const splitConfigurationData = {
    description: "Three percent variable",
    rules: [
      {
        paymentMethod: "ANY",
        shopperInteraction: "ANY",
        fundingSource: "ANY",
        currency: "ANY",
        splitLogic: {
          paymentFee: "deductFromLiableAccount",
          chargeback: "deductFromLiableAccount",
          chargebackCostAllocation: "deductFromLiableAccount",
          commission: {
            variablePercentage: 349
          },
          refund: "deductAccordingToSplitRatio",
          refundCostAllocation: "deductFromOneBalanceAccount"
        }
      }
    ]
  };

  const path = `/merchants/${merchantId}/splitConfigurations`;
  return handleAdyenRequest(pspClientConfig, 'POST', path, splitConfigurationData, 'create split configuration');
}

export async function createPaymentMethod(pspClientConfig: AdyenApiClientConfig, merchantId: string, businessLineId: string, paymentType: 'visa' | 'mc'): Promise<{ id: string }> {
  const paymentMethodData = {
    type: paymentType,
    currencies: ["USD"],
    countries: ["US"],
    businessLineId: businessLineId
  };

  const path = `/merchants/${merchantId}/paymentMethodSettings`;
  return handleAdyenRequest(pspClientConfig, 'POST', path, paymentMethodData, 'create payment method settings');
}

export async function createSweep(
    bpClientConfig: AdyenApiClientConfig,
    balanceAccountId: string,
    transferInstrumentId: string
): Promise<{ id: string }> {
  const sweep = {
    counterparty: {
      transferInstrumentId: transferInstrumentId
    },
    triggerAmount: {
      currency: 'USD',
      value: 0
    },
    currency: 'USD',
    priorities: ['regular', 'fast'],
    category: 'bank',
    schedule: {
      type: 'daily'
    },
    type: 'push',
    status: 'active'
  };

  const path = `/balanceAccounts/${balanceAccountId}/sweeps`;
  return handleAdyenRequest(bpClientConfig, 'POST', path, sweep, 'create sweep configuration')
}

export async function mapIndividualToSoleProprietorship(
    lemClientConfig: AdyenApiClientConfig,
    individualLegalEntityId: string,
    soleProprietorshipLegalEntityId: string
): Promise<void> {
  const path = `/legalEntities/${individualLegalEntityId}`;
  const payload = {
    entityAssociations: [
      {
        type: "soleProprietorship",
        "legalEntityId": soleProprietorshipLegalEntityId
      }
    ]
  };

  return handleAdyenRequest(lemClientConfig, 'PATCH', path, payload, 'map individual to sole proprietorship');
}

export async function createSoleProprietorshipLegalEntity(
    lemClientConfig: AdyenApiClientConfig,
    individualData: any
): Promise<{ id: string }> {
  const soleProprietorshipData = {
    type: "soleProprietorship",
    reference: `${individualData.reference}_sp`,
    soleProprietorship: {
      countryOfGoverningLaw: individualData?.individual?.residentialAddress?.country || "US",
      name: `${individualData?.individual?.name?.firstName} ${individualData?.individual?.name?.lastName}`.trim(),
      registeredAddress: {
        city: individualData?.individual?.residentialAddress?.city,
        country: individualData.individual?.residentialAddress?.country || "US",
        postalCode: individualData.individual?.residentialAddress?.postalCode || "",
        stateOrProvince: individualData.individual?.residentialAddress?.stateOrProvince || "",
        street: individualData.individual?.residentialAddress?.street || ""
      }
    }
  };

  const path = '/legalEntities';
  return handleAdyenRequest(lemClientConfig, 'POST', path, soleProprietorshipData, 'create sole proprietorship legal entity');
}