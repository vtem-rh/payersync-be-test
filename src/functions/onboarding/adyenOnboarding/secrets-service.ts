import { SecretsManagerClient, GetSecretValueCommand } from '@aws-sdk/client-secrets-manager';
import { captureAWSv3Client } from 'aws-xray-sdk-core';

const secretsManager = captureAWSv3Client(new SecretsManagerClient({}));

export async function getApiKey(secretName: string): Promise<string> {
  if (!secretName) {
    throw new Error(`Secret name is not defined.`);
  }
  const command = new GetSecretValueCommand({ SecretId: secretName });
  try {
    const result = await secretsManager.send(command);
    const secretString = result.SecretString;
    if (!secretString) {
      throw new Error(`API key not found in Secrets Manager for ${secretName}`);
    }
    
    // Parse the JSON secret string to extract the apiKey
    const secretData = JSON.parse(secretString);
    const apiKey = secretData.apiKey;
    if (!apiKey) {
      throw new Error(`API key not found in secret data for ${secretName}`);
    }
    
    return apiKey;
  } catch (err: any) {
    if (err.name === 'ResourceNotFoundException' || err.Code === 'ResourceNotFoundException') {
      console.error(`Secret not found: ${secretName}`, err);
      throw new Error(`Secret not found: ${secretName}`);
    }
    console.error(`Error fetching secret: ${secretName}`, err);
    throw err;
  }
}

export interface AdyenApiKeys {
  lemApiKey: string;
  bpApiKey: string;
  pspApiKey: string;
}

export async function getAdyenApiKeys(): Promise<AdyenApiKeys> {
  const lemSecretName = process.env.ADYEN_LEM_SECRET_NAME;
  const bpSecretName = process.env.ADYEN_BP_SECRET_NAME;
  const pspSecretName = process.env.ADYEN_PSP_SECRET_NAME;
  
  if (!lemSecretName || !bpSecretName || !pspSecretName) {
    throw new Error('Missing required environment variables for Adyen API key secret names');
  }
  
  const [lemApiKey, bpApiKey, pspApiKey] = await Promise.all([
    getApiKey(lemSecretName),
    getApiKey(bpSecretName),
    getApiKey(pspSecretName)
  ]);
  
  return { lemApiKey, bpApiKey, pspApiKey };
}