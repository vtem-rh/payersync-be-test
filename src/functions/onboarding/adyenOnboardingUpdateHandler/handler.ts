import { DynamoDBClient, UpdateItemCommand } from '@aws-sdk/client-dynamodb';
import { SNSEvent } from 'aws-lambda';
import { getTableName } from '../../shared/config-helpers';

const dynamo = new DynamoDBClient({});
const tableName = getTableName();

export const handler = async (event: SNSEvent) => {
  console.log('Handler started', JSON.stringify(event));
  for (const record of event.Records) {
    let userId: string | undefined;
    console.log('SNS Record:', JSON.stringify(record));
    console.log('SNS Message:', record.Sns.Message);
    try {
      const payload = JSON.parse(record.Sns.Message);
      console.log('Received SNS payload:', JSON.stringify(payload));
      userId = payload.userId;
      const adyenData = payload.adyenData;
      if (!userId) {
        console.error('Payload missing userId, cannot update DynamoDB', payload);
        throw new Error('500: Missing userId in payload');
      }
      if (adyenData === undefined) {
        console.error('Payload missing adyenData, nothing to update', payload);
        throw new Error('500: Missing adyenData in payload');
      }
      await dynamo.send(new UpdateItemCommand({
        TableName: tableName,
        Key: { userId: { S: userId } },
        UpdateExpression: 'SET adyenData = :adyenData, updatedAt = :updatedAt',
        ConditionExpression: 'attribute_exists(userId)',
        ExpressionAttributeValues: {
          ':adyenData': { S: JSON.stringify(adyenData) },
          ':updatedAt': { S: new Date().toISOString() },
        },
      }));
      console.log(`Updated adyenData for userId ${userId}`);
    } catch (err: any) {
      if (err.name === 'ConditionalCheckFailedException') {
        console.error(`DynamoDB update failed: userId ${userId} does not exist.`);
      } else {
        console.error('Failed to process SNS record', err, record);
        throw err;
      }
    }
  }
}; 