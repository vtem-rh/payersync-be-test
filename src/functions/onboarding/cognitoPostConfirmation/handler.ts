import { SNSClient, PublishCommand } from '@aws-sdk/client-sns';
import { captureAWSv3Client } from 'aws-xray-sdk-core';

interface CognitoEvent {
  userName: string;
  request: { userAttributes: { email: string; given_name?: string; family_name?: string } };
  region: string;
  triggerSource: string;
}

const snsClient = captureAWSv3Client(
  new SNSClient({
    region: process.env.AWS_REGION,
  })
);
const topicArn = process.env.LEAD_EVENT_TOPIC_ARN;

export const handler = async (event: CognitoEvent) => {
  const sanitizedEvent = {
    region: event.region,
    triggerSource: event.triggerSource,
  };
  console.log('Received Cognito event:', JSON.stringify(sanitizedEvent, null, 2));

  if (!topicArn) {
    const error = new Error('LEAD_EVENT_TOPIC_ARN environment variable is not set');
    console.error(error);
    throw error;
  }

  const { email, given_name, family_name } = event.request.userAttributes;
  const message = {
    userName: event.userName,
    email,
    givenName: given_name,
    familyName: family_name,
    eventType: 'USER_REGISTERED',
    timestamp: new Date().toISOString(),
    triggerSource: event.triggerSource,
  };

  try {
    console.log('Publishing message to SNS:', JSON.stringify(message, null, 2));
    console.log('Topic ARN:', topicArn);

    const result = await snsClient.send(
      new PublishCommand({
        TopicArn: topicArn,
        Message: JSON.stringify(message),
        Subject: 'Lead Event: User Registered',
      })
    );

    console.log('Successfully published to SNS. MessageId:', result.MessageId);
  } catch (error) {
    console.error('Failed to publish lead event to SNS:', error);
    throw error; // Re-throw to mark the Lambda execution as failed
  }

  return event;
};
