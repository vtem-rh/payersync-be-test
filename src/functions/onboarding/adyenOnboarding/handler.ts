import { APIGatewayProxyResult } from 'aws-lambda';
import { getAdyenApiKeys } from './secrets-service';
import * as AdyenApi from './adyen-api';

class ValidationError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'ValidationError';
  }
}

const returnError = (statusCode: number, message: string, error?: any) => {
  console.error(message, error);
  return {
    statusCode,
    headers: { 'Content-Type': 'application/json', 'Access-Control-Allow-Origin': '*' },
    body: JSON.stringify({ message, error: error ? error.message : undefined }),
  };
};

export const handler = async (event: any): Promise<APIGatewayProxyResult> => {
  try {
    if (!event.body) {
      return returnError(400, 'Missing request body');
    }
    const { merchantData } = JSON.parse(event.body);

    let lemApiKey: string | undefined;
    let bpApiKey: string | undefined;
    let pspApiKey: string | undefined;

    if (process.env.NODE_ENV === 'test') {
      console.log('Running in local test environment, decoding Base64 API keys from .env file...');
      const lemApiKeyBase64 = process.env.ADYEN_LEM_API_KEY_PARAM;
      const bpApiKeyBase64 = process.env.ADYEN_BP_API_KEY_PARAM;
      const pspApiKeyBase64 = process.env.ADYEN_PSP_API_KEY_PARAM;

      if (!lemApiKeyBase64 || !bpApiKeyBase64 || !pspApiKeyBase64) {
        console.log('Base64 encoded Adyen API keys are missing. Using mock API keys for tests...');
        lemApiKey = 'test_lem_api_key';
        bpApiKey = 'test_bp_api_key';
        pspApiKey = 'test_psp_api_key';
      } else {
        lemApiKey = Buffer.from(lemApiKeyBase64, 'base64').toString('ascii');
        bpApiKey = Buffer.from(bpApiKeyBase64, 'base64').toString('ascii');
        pspApiKey = Buffer.from(pspApiKeyBase64, 'base64').toString('ascii');
      }

    } else {
      console.log('Running in deployed environment, fetching API keys from Secrets Manager...');
      const keys = await getAdyenApiKeys();
      lemApiKey = keys.lemApiKey;
      bpApiKey = keys.bpApiKey;
      pspApiKey = keys.pspApiKey;
    }

    // Use mock API URLs if the actual ones are not available
    let bpApiUrl = process.env.ADYEN_BP_API_URL;
    let managementApiUrl = process.env.ADYEN_MANAGEMENT_API_URL;
    let lemApiUrl = process.env.ADYEN_LEM_API_URL;

    if (!bpApiUrl || !managementApiUrl || !lemApiUrl) {
      if (process.env.NODE_ENV === 'test') {
        console.log('Adyen API URLs are missing. Using mock URLs for tests...');
        bpApiUrl = 'https://127.0.0.1';
        managementApiUrl = 'https://127.0.0.1';
        lemApiUrl = 'https://127.0.0.1';
      } else {
        return returnError(500, 'One or more Adyen API URLs are missing. Check environment variables ADYEN_BP_API_URL, ADYEN_MANAGEMENT_API_URL, and ADYEN_LEM_API_URL.');
      }
    }

    if (!lemApiKey || !bpApiKey || !pspApiKey) {
      return returnError(500, 'One or more Adyen API keys are missing. For local tests, check ADYEN_*_API_KEY in your .env file. For deployed environments, check Secrets Manager.');
    }

    const adyenMerchantAccount = process.env.ADYEN_MERCHANT_ACCOUNT;
    if (!adyenMerchantAccount) {
      return returnError(500, 'ADYEN_MERCHANT_ACCOUNT environment variable is not set.');
    }

    const apiClients = AdyenApi.createAdyenApiClients(lemApiKey, bpApiKey, pspApiKey, bpApiUrl, managementApiUrl, lemApiUrl);

    // Build the legal entity payload from merchantData
    const { id: legalEntityId } = await AdyenApi.createLegalEntity(apiClients.lem, merchantData.legalEntity);

    // COMMENTED OUT - Missing required data in current payload
    // const { id: accountHolderId } = await AdyenApi.createAccountHolder(apiClients.bp, merchantData.accountHolder, legalEntityId);
    // await AdyenApi.createBalanceAccount(apiClients.bp, merchantData.balanceAccount, accountHolderId);

    // COMMENTED OUT - Missing required data in current payload
    // // Ensure businessLine.webData is always an array
    // if (merchantData.businessLine) {
    //   if (merchantData.businessLine.webData !== undefined) {
    //     if (!Array.isArray(merchantData.businessLine.webData)) {
    //       merchantData.businessLine.webData = [merchantData.businessLine.webData];
    //     }
    //   } else {
    //     merchantData.businessLine.webData = [];
    //   }
    // }
    // const { id: businessLineId } = await AdyenApi.createBusinessLine(apiClients.lem, merchantData.businessLine, legalEntityId);

    // COMMENTED OUT - Missing businessLineId
    // await AdyenApi.createStore(apiClients.psp, merchantData.store, businessLineId, adyenMerchantAccount);

    // Create a minimal hostedOnboarding object since it's missing from payload
    const hostedOnboarding = merchantData.hostedOnboarding || {
      // Add minimal required fields - you may need to adjust based on Adyen requirements
      platformName: "PayerSync",
      // Add other required fields as needed
    };

    const { url } = await AdyenApi.createOnboardingLink(apiClients.lem, hostedOnboarding, legalEntityId);

    const successfulOnboardingData = {
      ...merchantData,
      adyenData: {
        legalEntityId,
        // accountHolderId,  // Commented out since we're not creating it
        // businessLineId,   // Commented out since we're not creating it
        url,
      }
    };

    console.log('Onboarding successful', successfulOnboardingData);
    return {
      statusCode: 200,
      headers: { 'Content-Type': 'application/json', 'Access-Control-Allow-Origin': '*' },
      body: JSON.stringify(successfulOnboardingData),
    };
  } catch (error: any) {
    // Explicitly log all errors for observability
    console.error('Error in Adyen onboarding Lambda:', error);
    if (error instanceof ValidationError) {
      return returnError(400, error.message);
    }
    return returnError(500, 'Failed to process Adyen onboarding', error);
  }
};
