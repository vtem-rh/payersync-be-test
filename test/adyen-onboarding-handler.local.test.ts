// This test file is for local-only tests that may call real Adyen endpoints.
// Do NOT run this file in CI/CD.
if (process.env.CI) {
  // eslint-disable-next-line no-undef
  describe.skip('Adyen Onboarding Handler (Local Only)', () => {
    it('skipped in CI', () => {});
  });
  // Exit early so no other tests run
  // @ts-ignore
  return;
}
import { handler } from '../src/functions/onboarding/adyenOnboarding/handler';
import * as AdyenApi from '../src/functions/onboarding/adyenOnboarding/adyen-api';
import * as fs from 'fs';
import * as path from 'path';

// Mock the Adyen API module
jest.mock('../src/functions/onboarding/adyenOnboarding/adyen-api');

const mockedAdyenApi = AdyenApi as jest.Mocked<typeof AdyenApi>;

describe('Adyen Onboarding Handler (Local Only)', () => {
  let merchantData: any;

  beforeEach(() => {
    jest.resetAllMocks();
    const payloadPath = path.join(__dirname, 'events/adyen-onboarding-payload.json');
    const payload = fs.readFileSync(payloadPath, 'utf-8');
    merchantData = JSON.parse(payload).merchantData;
    process.env.ADYEN_MERCHANT_ACCOUNT = 'RectangleHealthCOM';

    mockedAdyenApi.createLegalEntity.mockResolvedValue({ id: 'LE123' });
    mockedAdyenApi.createAccountHolder.mockResolvedValue({ id: 'AH123' });
    mockedAdyenApi.createBalanceAccount.mockResolvedValue({ id: 'BA123' });
    mockedAdyenApi.createBusinessLine.mockResolvedValue({ id: 'BL123' });
    mockedAdyenApi.createStore.mockResolvedValue({ id: 'ST123' });
    mockedAdyenApi.createOnboardingLink.mockResolvedValue({ url: 'https://onboarding.adyen.com/link/123' });
    mockedAdyenApi.createAdyenApiClients.mockReturnValue({
      lem: 'mockedLemClient',
      bp: 'mockedBpClient',
      psp: 'mockedPspClient',
    } as any);
  });

  it('should process the onboarding flow successfully and return an onboarding link', async () => {
    // Arrange
    const event = {
      body: JSON.stringify({ merchantData, userId: 'test-user-id' }),
    };

    // Act
    const result = await handler(event);

    // Assert
    expect(merchantData).toMatchSnapshot('merchantData-payload');
    expect(result.statusCode).toBe(200);
    const responseBody = JSON.parse(result.body);
    expect(responseBody.adyenData).toHaveProperty('url');
    expect(responseBody.adyenData.url).toBe('https://onboarding.adyen.com/link/123');
    expect(responseBody).toMatchObject({
      adyenData: expect.any(Object),
    });
    expect(mockedAdyenApi.createLegalEntity).toHaveBeenCalledWith('mockedLemClient', expect.any(Object));
    expect(mockedAdyenApi.createAccountHolder).toHaveBeenCalledWith('mockedBpClient', merchantData.accountHolder, 'LE123');
    expect(mockedAdyenApi.createBusinessLine).toHaveBeenCalledWith('mockedLemClient', merchantData.businessLine, 'LE123');
    expect(mockedAdyenApi.createBalanceAccount).toHaveBeenCalledWith('mockedBpClient', merchantData.balanceAccount, 'AH123');
    expect(mockedAdyenApi.createStore).toHaveBeenCalledWith(
      'mockedPspClient',
      merchantData.store,
      'BL123',
      'RectangleHealthCOM'
    );
    expect(mockedAdyenApi.createOnboardingLink).toHaveBeenCalledWith('mockedLemClient', 'LE123');
  });
}); 