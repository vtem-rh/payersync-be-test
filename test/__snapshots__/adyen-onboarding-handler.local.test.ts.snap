// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Adyen Onboarding Handler (Local Only) should process the onboarding flow successfully and return an onboarding link: merchantData-payload 1`] = `
{
  "accountHolder": {
    "description": "Test Merchant Account Holder",
    "reference": "test-merchant-001",
  },
  "balanceAccount": {
    "defaultCurrencyCode": "USD",
    "description": "USD Balance Account for Test Merchant",
  },
  "businessLine": {
    "industryCode": "339E",
    "salesChannels": [
      "eCommerce",
    ],
    "service": "paymentProcessing",
    "webData": [
      {
        "webAddress": "https://testmerchant.com",
      },
    ],
  },
  "hostedOnboarding": {
    "locale": "en-US",
    "redirectUrl": "https://testmerchant.com/onboarding-complete",
    "themeId": "ONB_THEME_001",
  },
  "legalEntity": {
    "organization": {
      "legalName": "Test Merchant Inc.",
      "registeredAddress": {
        "city": "Greenacres",
        "country": "US",
        "postalCode": "33463",
        "stateOrProvince": "FL",
        "street": "123 Test Street",
      },
      "type": "privateCompany",
    },
    "type": "organization",
  },
  "store": [
    {
      "address": {
        "city": "Springfield",
        "country": "US",
        "line1": "200 Main Street",
        "line2": "Building 5A",
        "line3": "Suite 3",
        "postalCode": "20250",
        "stateOrProvince": "NY",
      },
      "description": "Test Merchant Online Store",
      "phoneNumber": "+13123456789",
      "reference": "test-merchant-store-1",
      "shopperStatement": "TestMerchant",
    },
  ],
}
`;
